README.md
# zeroml


[![pypi](https://img.shields.io/pypi/v/zeroml.svg)](https://pypi.org/project/zeroml/)
[![python](https://img.shields.io/pypi/pyversions/zeroml.svg)](https://pypi.org/project/zeroml/)
[![Build Status](https://github.com/i008/zeroml/actions/workflows/dev.yml/badge.svg)](https://github.com/i008/zeroml/actions/workflows/dev.yml)
[![codecov](https://codecov.io/gh/i008/zeroml/branch/main/graphs/badge.svg)](https://codecov.io/github/i008/zeroml)



zeroml


CHANGELOG.md
# Changelog

## 0.1.0 (2022-02-17)

* First release on PyPI.

## 0.2.1 (2022-02-17)

* Test


tests/test_io.py
import numpy as np
import PIL.Image

from zeroml.io.image import save_array_image, save_pil


def test_save_pillow_image_random_name(random_pil_image, tmp_path):

    fp = save_pil(random_pil_image, tmp_path)
    image = PIL.Image.open(fp)

    assert image.size == random_pil_image.size


def test_save_numpy_image(tmp_path):
    image = np.random.randn(224, 224, 3).astype("uint8")
    fp = save_array_image(image, tmp_path)
    image = PIL.Image.open(fp)

    assert image.size == (224, 224)


tests/conftest.py
import numpy as np
import PIL.Image
import pytest


@pytest.fixture
def random_pil_image() -> PIL.Image:
    image = PIL.Image.fromarray(np.random.randn(224, 224, 3).astype("uint8"))
    return image


tests/test_dataloading_images.py
import uuid

import numpy as np
import pandas as pd
import PIL.Image
import pytest

from zeroml.dataloading.image.datasets import ImageClfAlbuDataset
from zeroml.dataloading.image.transforms import build_training_transform


@pytest.fixture
def create_test_dataframe(tmp_path):

    images = [PIL.Image.fromarray(np.ones((224, 224, 3)).astype("uint8")) for _ in range(10)]
    labels = [i for i in range(10)]
    fns = [str(uuid.uuid4()) + '.png' for _ in range(10)]
    print(fns)

    for image, f in zip(images, fns):
        image.save(tmp_path / f)

    df = pd.DataFrame({'labels': labels, 'file_names': fns})
    base_path = tmp_path
    return df, base_path


def test_clf_albu_dataset(create_test_dataframe):

    df, base_path = create_test_dataframe

    transform = build_training_transform(224, model=None, augment_level='noaug')

    ds = ImageClfAlbuDataset(
        df, base_path=base_path, transform=transform, file_column='file_names', label_column='labels'
    )

    image, label = ds[5]

    assert label == 5


zeroml/dataloading/image/transforms.py
from typing import List

import albumentations as A
from albumentations.pytorch.transforms import ToTensorV2
from timm.models.registry import _model_default_cfgs


def build_post_transform(model: str, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)):
    if model:
        cfg = _model_default_cfgs[model.split("timm/")[-1] if "timm/" in model else model]
        mean = cfg["mean"]
        std = cfg["std"]
        print("Using data config", cfg)

    return [A.Normalize(mean=mean, std=std), ToTensorV2()]


def build_pre_transform(size):
    pre_transform = [A.LongestMaxSize(size), A.PadIfNeeded(size, size, border_mode=0)]
    return pre_transform


def timm_inference_transform(model_name, image_size=224):
    post_transform = build_post_transform(model_name)
    pre_transform = [
        A.LongestMaxSize(image_size),
        A.PadIfNeeded(image_size, image_size, border_mode=0),
    ]

    return A.Compose([*pre_transform, *post_transform])


def get_augment(augment_level: str) -> List[A.BasicTransform]:
    if augment_level not in AUGMENTATIONS:
        raise ValueError(f"Augmentation strategy has to be one of {AUGMENTATIONS.keys()}")
    return AUGMENTATIONS[augment_level]


def build_visualize_transform(size, augment_level: str):
    pre_transform = build_pre_transform(size)
    augment_transform = get_augment(augment_level)

    return A.Compose([*pre_transform, *augment_transform])


def build_training_transform(size, model, augment_level: str) -> A.Compose:
    pre_transform = build_pre_transform(size)
    post_transform = build_post_transform(model)
    augment_transform = get_augment(augment_level)

    return A.Compose([*pre_transform, *augment_transform, *post_transform])


def build_inference_transform(model: str, size=224) -> A.Compose:
    pre_transform = build_pre_transform(size)
    post_transform = build_post_transform(model)
    return A.Compose([*pre_transform, *post_transform])


def build_eval_transform(model, size) -> A.Compose:
    pre_transform = build_pre_transform(size)
    post_transform = build_post_transform(model)

    return A.Compose([*pre_transform, *post_transform])


AUGMENTATIONS = {
    "hard_1": [
        A.RandomRotate90(),
        A.Flip(),
        A.Transpose(),
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=3, p=0.1),
            ],
            p=0.2,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=45, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
    ],
    "medium": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=3, p=0.1),
            ],
            p=0.2,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=45, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
    ],
    "medium3": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(blur_limit=15, p=1),
                A.MedianBlur(blur_limit=15, p=1),
                A.Blur(blur_limit=15, p=1),
            ],
            p=1,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.1, rotate_limit=15, p=0.2),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
    ],
    "medium2": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=3, p=0.1),
            ],
            p=0.5,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=15, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
        A.HorizontalFlip(p=0.33),
    ],
    "medium2_strong_blur": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                # A.MotionBlur(p=0.2),
                # A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=(9, 11), p=1),
            ],
            p=1,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=15, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
        A.HorizontalFlip(p=0.33),
    ],
    "medium4": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=7, p=0.1),
                A.Blur(blur_limit=7, p=0.1),
            ],
            p=0.5,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=15, p=0.2, border_mode=0),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
                A.ColorJitter(),
                A.ToGray(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=2,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
        A.HorizontalFlip(p=0.33),
    ],
    "easy_2": [A.RandomBrightnessContrast()],
    "noaug": [],
    "new": [
        A.IAAAdditiveGaussianNoise(p=0.5),
        A.IAASuperpixels(p=1),
        A.ImageCompression(p=1),
        A.IAAPerspective(p=1),
        A.RGBShift(r_shift_limit=50, b_shift_limit=50, g_shift_limit=50, p=1),
        A.Posterize(p=1, num_bits=3),
        A.IAAAdditiveGaussianNoise(p=1, loc=0, scale=(30, 50)),
    ],
    "kaggle": [
        # A.HorizontalFlip(p=0.5),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=7, p=0.1),
                A.Blur(blur_limit=7, p=0.1),
            ],
            p=0.5,
        ),
        A.HueSaturationValue(p=0.3),
        A.RGBShift(r_shift_limit=10, b_shift_limit=10, g_shift_limit=10, p=0.1),
        A.ImageCompression(quality_lower=50, quality_upper=100),
        A.ShiftScaleRotate(shift_limit=0.1, scale_limit=0.1, rotate_limit=10, border_mode=0, p=0.5),
    ],
}


zeroml/dataloading/image.py
from typing import List

import albumentations as A
from albumentations.pytorch.transforms import ToTensorV2
from timm.models.registry import _model_default_cfgs


def build_post_transform(model: str, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)):
    if model:

        cfg = _model_default_cfgs[model.split("timm/")[-1] if "timm/" in model else model]
        mean = cfg["mean"]
        std = cfg["std"]
        print("Using data config", cfg)

    return [A.Normalize(mean=mean, std=std), ToTensorV2()]


def build_pre_transform(size):
    pre_transform = [A.LongestMaxSize(size), A.PadIfNeeded(size, size, border_mode=0)]
    return pre_transform


def timm_inference_transform(model_name, image_size=224):
    post_transform = build_post_transform(model_name)
    pre_transform = [
        A.LongestMaxSize(image_size),
        A.PadIfNeeded(image_size, image_size, border_mode=0),
    ]

    return A.Compose([*pre_transform, *post_transform])


def get_augment(augment_level: str) -> List[A.BasicTransform]:
    if augment_level not in AUGMENTATIONS:
        raise ValueError(f"Augmentation strategy has to be one of {AUGMENTATIONS.keys()}")
    return AUGMENTATIONS[augment_level]


def build_visualize_transform(size, augment_level: str):
    pre_transform = build_pre_transform(size)
    augment_transform = get_augment(augment_level)

    return A.Compose([*pre_transform, *augment_transform])


def build_training_transform(size, model, augment_level: str) -> A.Compose:
    pre_transform = build_pre_transform(size)
    post_transform = build_post_transform(model)
    augment_transform = get_augment(augment_level)

    return A.Compose([*pre_transform, *augment_transform, *post_transform])


def build_inference_transform(model: str, size=224) -> A.Compose:
    pre_transform = build_pre_transform(size)
    post_transform = build_post_transform(model)
    return A.Compose([*pre_transform, *post_transform])


def build_eval_transform(model, size) -> A.Compose:
    pre_transform = build_pre_transform(size)
    post_transform = build_post_transform(model)

    return A.Compose([*pre_transform, *post_transform])


AUGMENTATIONS = {
    "hard_1": [
        A.RandomRotate90(),
        A.Flip(),
        A.Transpose(),
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=3, p=0.1),
            ],
            p=0.2,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=45, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
    ],
    "medium": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=3, p=0.1),
            ],
            p=0.2,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=45, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.2,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
    ],
    "medium3": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(blur_limit=15, p=1),
                A.MedianBlur(blur_limit=15, p=1),
                A.Blur(blur_limit=15, p=1),
            ],
            p=1,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.1, rotate_limit=15, p=0.2),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
    ],
    "medium2": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=3, p=0.1),
            ],
            p=0.5,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=15, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
        A.HorizontalFlip(p=0.33),
    ],
    "medium2_strong_blur": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                # A.MotionBlur(p=0.2),
                # A.MedianBlur(blur_limit=3, p=0.1),
                A.Blur(blur_limit=(9, 11), p=1),
            ],
            p=1,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=15, p=0.2),
        A.OneOf(
            [
                A.OpticalDistortion(p=0.3),
                A.GridDistortion(p=0.1),
                A.IAAPiecewiseAffine(p=0.3),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=1,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
        A.HorizontalFlip(p=0.33),
    ],
    "medium4": [
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=7, p=0.1),
                A.Blur(blur_limit=7, p=0.1),
            ],
            p=0.5,
        ),
        A.ShiftScaleRotate(shift_limit=0.0625, scale_limit=0.2, rotate_limit=15, p=0.2, border_mode=0),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
                A.ColorJitter(),
                A.ToGray(),
            ],
            p=0.3,
        ),
        A.HueSaturationValue(p=0.3),
        A.CoarseDropout(
            max_holes=2,
            max_height=100,
            max_width=50,
            p=0.66,
            min_holes=1,
            min_height=50,
            min_width=20,
        ),
        A.JpegCompression(),
        A.HorizontalFlip(p=0.33),
    ],
    "easy_2": [A.RandomBrightnessContrast()],
    "noaug": [],
    "new": [
        A.IAAAdditiveGaussianNoise(p=0.5),
        A.IAASuperpixels(p=1),
        A.ImageCompression(p=1),
        A.IAAPerspective(p=1),
        A.RGBShift(r_shift_limit=50, b_shift_limit=50, g_shift_limit=50, p=1),
        A.Posterize(p=1, num_bits=3),
        A.IAAAdditiveGaussianNoise(p=1, loc=0, scale=(30, 50)),
    ],
    "kaggle": [
        # A.HorizontalFlip(p=0.5),
        A.OneOf(
            [
                A.CLAHE(clip_limit=2),
                A.IAASharpen(),
                A.IAAEmboss(),
                A.RandomBrightnessContrast(),
            ],
            p=0.3,
        ),
        A.OneOf(
            [
                A.IAAAdditiveGaussianNoise(),
                A.GaussNoise(),
            ],
            p=0.5,
        ),
        A.OneOf(
            [
                A.MotionBlur(p=0.2),
                A.MedianBlur(blur_limit=7, p=0.1),
                A.Blur(blur_limit=7, p=0.1),
            ],
            p=0.5,
        ),
        A.HueSaturationValue(p=0.3),
        A.RGBShift(r_shift_limit=10, b_shift_limit=10, g_shift_limit=10, p=0.1),
        A.ImageCompression(quality_lower=50, quality_upper=100),
        A.ShiftScaleRotate(shift_limit=0.1, scale_limit=0.1, rotate_limit=10, border_mode=0, p=0.5),
    ],
}
